agents/aggregator/aggregator_agent.py

```diff
@@ -18,463 +13,511 @@ class Provider:
-                self.log.info("Scheduled remove for provider {} ({})"
-                              .format(feed['address'], session_id))
-                prov.remove = True
-
-            if action in ['added', 'status']:
-                if prov is not None:
-                    self.log.warn("Provider {} ({}) already exists.".format(feed['address'], session_id))
-                    return
-
-                prov_id = self.hksess.add_provider(
-                    description="{}".format(feed['address'])
-                )
-                self.providers[prov_id] = Provider(feed, prov_id)
-                self.prov_ids[(feed["address"], session_id)] = prov_id
-                self.log.info("Added provider {} (session_id: {}) with id {}"
-                              .format(feed['address'], session_id, prov_id))
-                self.should_write_status = True
-                self.agent.subscribe_to_feed(agent_address,
-                                             feed['feed_name'],
-                                             self._data_handler)
-
-    def add_feed(self, session, params={}):
+    def remove_provider(self, prov):
         """
-        Task to subscribe to a feed.
+        Writes remaining provider data to frame and removes provider.
 
-        Arguments:
-            address (str):
-                Full address of the feed
-            frame_length:
-                Time before frames should be written to disk
+        Args:
+            prov (Provider):
+                provider object that should be removed.
         """
-        agent_addr, feed_name = params['address'].split('.feeds.')
-        prov_id = self.hksess.add_provider(
-            description="{}".format(params['address'])
-        )
-
-        x = {
-            'session_id': -1,
-            'address': params['address'],
-            'frame_length': params['frame_length'],
-        }
+        pid = prov.prov_id
+        addr, sessid = prov.address, prov.sessid
 
-        self.providers[prov_id] = Provider(x, prov_id)
-        self.agent.subscribe_to_feed(agent_addr,
-                                     feed_name,
-                                     self._data_handler)
-        self.should_write_status = True
-        return True, "Feed added"
+        if not prov.empty():
+            self.frame_queue.put(prov.to_frame(self.hksess, clear=False))
 
-    def initialize(self, session, params=None):
-        """TASK: Subscribes to `agent_activity` feed and has registry dump
-        info on all active agents.  Optionally starts the "record"
-        Process.
-
-        ``params`` is a dict with the following keys:
+        self.hksess.remove_provider(pid)
+        del self.providers[pid]
+        del self.pids[(addr, sessid)]
+        self.write_status()
 
-        - 'start_record' (bool, optional): Default is False.  If True,
-            start the "record" Process after performing registration.
+    def write_data(self, data, pid):
```

Comment:
Within this class, "write" is used to describe each of:
- transfering data into a Provider storage object.
- transfering data from Providers to our queue.
- transfering data from queue to file.

So maybe "write_data" should be more descriptive, like the other write methods -- "write_data_to_provider"?  My preference would be to use different verbs though... store/enqueue/write?